net: Filter addrman during address selection via AddrPolicy to avoid underfill

waketraindev opened the pull request:
This PR introduces `AddrMan::AddrPolicy`, a predicate that allows callers to
exclude addresses during selection. The policy returns true to skip a given
entry and is evaluated while holding `AddrMan::cs`.

The mechanism is used in `CConnman::GetAddressesUnsafe()` to filter out
banned and discouraged peers during address selection instead of removing
them afterward. This prevents P2P responses (`GETADDR`) and RPCs (`getnodeaddresses)` from
returning fewer results than requested when large portions of the address space are filtered.

Additional improvements:
- Logs the number of filtered entries for diagnostics
- Avoids redundant filtering passes in higher-level callers
- Keeps locking behavior unchanged (the policy runs under `AddrMan::cs`)

**Testing**
- Repro: ban or discourage a large set of addresses, then call
  `getnodeaddresses 1`,  previously underfilled.
- After this change, the RPC returns up to the requested count as expected.

No behavior change for unaffected callers.

mzumsande commented on src/net.cpp:
```diff
@@ -3527,13 +3527,11 @@ CConnman::~CConnman()
-                        [this](const CAddress& addr){return m_banman->IsDiscouraged(addr) || m_banman->IsBanned(addr);}),
-                        addresses.end());
+        const AddrMan::AddrPolicy policy = [this](const CAddress& addr) { return m_banman->IsDiscouraged(addr) || m_banman->IsBanned(addr); };
+        return addrman.GetAddr(max_addresses, max_pct, network, filtered, policy);
```
The main funcion of this is p2p  (answer GetAddr requests), RPCs like `getnodeaddresses` are of secondary importance. Since the p2p use is also affected by this, the PR description should be changed to reflect this.

mzumsande reviewed (commented):
Concept ACK

The main function of this is p2p  (answering GetAddr requests from peers), RPCs like `getnodeaddresses` are of secondary importance. 
Since the p2p use is also affected by this, the PR description should be changed to reflect this. 

That said, it seems good to always try to return 1000 addresses in a GetAddr answer, so that peers don't know whether we have a long list of banned addresses or not.

waketraindev commented on src/addrman.cpp:
```diff
@@ -839,14 +840,26 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
+            continue;
+        }
+
+        // Filter by policy
```
Will do, looking into coverege *and sonarcloud* right now

brunoerg commented on src/addrman.cpp:
```diff
@@ -839,14 +840,26 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
+        // Filter by policy
+        if (policy && policy(ai)) {
+            ++filtered_addresses;
+            continue;
```
Thanks for adding test coverage. By doing mutation testing for this PR, I noticed that the following mutant has not been killed (it means tests were not able to detect this change):

```diff
diff --git a/src/addrman.cpp b/src/addrman.cpp
index c06fbb0673..4cc7073c3b 100644
--- a/src/addrman.cpp
+++ b/src/addrman.cpp
@@ -854,7 +854,7 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
         // Filter by policy
         if (policy && policy(ai)) {
             ++filtered_addresses;
-            continue;
+            break;
         }
```

This is because you're checking that the addresses that are returned when filtering by port are - de facto - according to the policy. However, you could also check that all the proper addresses are being returned, I think just adding the following check would be enough:

```diff
diff --git a/src/test/addrman_tests.cpp b/src/test/addrman_tests.cpp
index 987249104d..3073656ed5 100644
--- a/src/test/addrman_tests.cpp
+++ b/src/test/addrman_tests.cpp
@@ -421,6 +421,7 @@ BOOST_AUTO_TEST_CASE(addrman_getaddr)
     BOOST_CHECK_EQUAL(vAddrNoPolicy.size(), addedAddresses);
     BOOST_CHECK_EQUAL(vAddrPolicyNoSkip.size(), vAddrNoPolicy.size());
     BOOST_CHECK_EQUAL(vAddrPolicySkipAll.size(), 0);
+    BOOST_CHECK_EQUAL(vAddrPolicyPort.size(), 4U);
     BOOST_CHECK(std::none_of(vAddrPolicyPort.begin(), vAddrPolicyPort.end(), policyPortPolicy));
```

waketraindev commented:
Added test verifying only expected addresses are returned and filtered ones excluded.

frankomosh commented on src/net.cpp:
```diff
@@ -3527,13 +3527,10 @@ CConnman::~CConnman()
-                        addresses.end());
-    }
-    return addresses;
+        const AddrMan::AddrPolicy policy = [this](const CAddress& addr) {
```
Is there a way/test to verify that the `BanMan` methods here are safe to call while `AddrMan`'s lock is held? 

Can see a comment in `addrman.h` says the policy "must not acquire locks that can conflict with `AddrMan::cs`". So it would make sense to ensure deadlock scenario is not accidentally created.

waketraindev replied:
Updated comments for more clarity
in addrman.h:
```cpp
    /** Predicate used to exclude addresses during selection.
     *  Return true to skip the given address.
     *
     *  Runs while holding AddrMan::cs, so it must be non-blocking, must not
     *  attempt to reacquire AddrMan::cs, and must preserve lock ordering to
     *  avoid deadlocks.
     */
```
in net.cpp -> GetAddressesUnsafe()
```cpp
        // This runs under AddrMan::cs, and BanMan checks take m_banned_mutex.
        // The lock order here is AddrMan::cs -> m_banned_mutex.
```

frankomosh reviewed (commented):
Concept ACK

waketraindev commented on src/addrman.cpp:
```diff
@@ -839,14 +840,26 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
         // Filter by network (optional)
-        if (network != std::nullopt && ai.GetNetClass() != network) continue;
+        if (network != std::nullopt && ai.GetNetClass() != network) {
+            ++filtered_addresses;
```
Yes; Currently all skipped addresses in `AddrMan::GetAddr` are counted as `filtered`; If you'd prefer (or find it useful) separate counters for each reason (Network, Quality or Policy), I can add that.

frankomosh replied:
Maybe separate would be clearer for diagnostics. Up to you though,.. if you think the current logging is clear enough then that's fine too.

waketraindev replied:
Went with separate counters and also added the selected network to the log output for clarity.
Example log entry:
```
[addrman] GetAddr returned 62823 random addresses from ipv4; 1104 filtered (0 network, 0 quality, 1104 policy)
[addrman] GetAddr returned 0 random addresses from ipv6; 63927 filtered (63927 network, 0 quality, 0 policy)
[addrman] GetAddr returned 0 random addresses from onion; 63927 filtered (66927 network, 0 quality, 0 policy)
```

mzumsande reviewed (commented):
Code Review ACK fb3efdb3d267924266134f63b1ba6ddda0b6ce81

As far as I can see, the reason for creating and passing an `AddrPolicy` is that the alternative (having `addrman` reach into `banman` to request if an addr is banned) would introduce coupling between the modules, while the current approach only introduces an implied lock order, which is somewhat of a potential footgun - but I don't really see a better solution. 
For example, we probably wouldn't want to remove entries from addrman when they are banned, because bans are often temporary.

fjahr commented on src/addrman.cpp:
```diff
@@ -827,6 +828,7 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
     const auto now{Now<NodeSeconds>()};
     std::vector<CAddress> addresses;
     addresses.reserve(nNodes);
+    size_t filtered_addr_net = 0, filtered_addr_quality = 0, filtered_addr_policy = 0;
```
A general problem with the separate tracking like this is that it's order dependent and so the values may actually be somewhat misleading. I.e. if there are some addrs filtered by policy but they are already caught by the net filter they won't be included here. Doing this in a cleaner way may be overengineering but I wanted to mention that this isn't ideal.

fjahr replied:
Adding to that: the non-optional filter is now in between two optional filters. Feels to me like the optional checks should go after the mandatory but maybe there are arguments for doing it the other way around.

fjahr replied:
If my suggestion from the top level comment is taken this is kind of obsolete but if not i think rather than trying to count each filter individually I would rather just log the total filtered count and which filters were active. I think that's simpler and least confusing.

fjahr commented on src/addrman.cpp:
```diff
@@ -839,14 +841,31 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
     }
-    LogDebug(BCLog::ADDRMAN, "GetAddr returned %d random addresses\n", addresses.size());
+    LogDebug(BCLog::ADDRMAN, "GetAddr returned %zu random addresses from %s; %zu filtered "
+                             "(%zu network, %zu quality, %zu policy)\n",
```
Would be generally nice if the options that weren't used are also not logged.

fjahr reviewed (commented):
While leaving some nit-ish comments I have realized that if we want to add such a generalized concept for filtering, we should implement the current filtering with it as well. E.g. `network` and `filtered` params go away and we only have `AddrPolicy` as param to `GetAddr`. Then use `AddrPolicy` to implement what `network` and `filtered` were doing. That's a bigger refactor but having this current mix with this redundancy is kind of ugly and I don't really think it's worth it alone.

fjahr commented:
> network and filtered serve distinct, standard purposes (network filters by network, filtered disables quality filtering), while AddrPolicy is specifically for custom filters

What distinguishes a "standard" from a "custom" filter? Is that documented somewhere? And if we need to add another filter that is "standard" we would add another param to the function rather than using `AddrPolicy`?

> that skip entries without reducing the returned count

Huh? If they don't reduce the count how are they filtering anything? This is not what the code is doing, the code is filtering and reducing the returned count with it as well of course. Also demonstrated by the tests. Am I misunderstanding what you mean with count?

> Using AddrPolicy also avoids coupling AddrMan to other modules (like BanMan) for filtering.

I don't see how that is an argument against using it for network and quality filters.

waketraindev commented:
> > network and filtered serve distinct, standard purposes (network filters by network, filtered disables quality filtering), while AddrPolicy is specifically for custom filters
> 
> What distinguishes a "standard" from a "custom" filter? Is that documented somewhere? And if we need to add another filter that is "standard" we would add another param to the function rather than using `AddrPolicy`?

By "standard" I mean filters that are intrinsic to `AddrMan` and operate on `AddrInfo` (for example network classification and quality filtering via `ai.IsTerrible(now)`), and whose logic does not depend on external subsystems. By "custom" I mean filters that depend on external state or policy decisions outside of `AddrMan` (for example `BanMan`), which is what `AddrPolicy` is intended to support.
If a future filter is intrinsic to `AddrMan` and depends on `AddrInfo`, adding it as a separate parameter would be consistent with the existing design. If it depends on external state, `AddrPolicy` is the appropriate mechanism.

> > that skip entries without reducing the returned count
> 
> Huh? If they don't reduce the count how are they filtering anything? This is not what the code is doing, the code is filtering and reducing the returned count with it as well of course. Also demonstrated by the tests. Am I misunderstanding what you mean with count?

Candidates are filtered and skipped during selection. The change here is *when* filtering happens. Previously, addresses were selected up to the target count and then banned or discouraged addresses were removed afterward, which caused the final `CConnman::GetAddressesUnsafe` result to be underfilled. With the current approach, filtering happens during selection, and the loop continues until the requested number of unfiltered addresses is reached or the address space is exhausted. So filtering still occurs, but it no longer causes `CConnman::GetAddressesUnsafe` to be underfilled relative to the request.

> > Using AddrPolicy also avoids coupling AddrMan to other modules (like BanMan) for filtering.
> 
> I don't see how that is an argument against using it for network and quality filters.

The point is not that `AddrPolicy` cannot express network or quality filters, but that doing so would conflate concerns and duplicate logic. Network and quality filtering are `AddrMan` responsibilities and require access to `AddrInfo`. Expressing them via `AddrPolicy` would either broaden the policy contract or require reimplementing intrinsic `AddrMan` logic in call sites, without addressing a concrete issue in this PR. This change keeps the PR scoped to fixing underfilling and avoids introducing additional coupling or refactors.

---

To keep this PR focused: its purpose is to fix the underfilling issue by moving ban and discourage filtering into the selection phase. The current design achieves that with minimal surface area change and without refactoring existing `AddrMan` responsibilities or introducing new abstractions. Broader refactors (for example unifying all filters behind `AddrPolicy` or introducing builders/factories) can be explored separately if there is concrete motivation, but are out of scope for this change. Unless there are functional issues with the current approach, I intend to keep the implementation as is.

fjahr commented on src/addrman.cpp:
```diff
@@ -23,6 +23,7 @@
 
 #include <cmath>
 #include <optional>
+#include <netbase.h>
```
nit: ordering

fjahr reviewed (commented):
> By "standard" I mean filters that are intrinsic to `AddrMan` and operate on `AddrInfo` (for example network classification and quality filtering via `ai.IsTerrible(now)`), and whose logic does not depend on external subsystems. By "custom" I mean filters that depend on external state or policy decisions outside of `AddrMan` (for example `BanMan`), which is what `AddrPolicy` is intended to support.
> If a future filter is intrinsic to `AddrMan` and depends on `AddrInfo`, adding it as a separate parameter would be consistent with the existing design. If it depends on external state, `AddrPolicy` is the appropriate mechanism.

The docs don't mention any of this but instead the naming `AddrPolicy` and the docs suggest that it's a generalized concept and could be used for any kind of filtering. I still think there is no reason to restrict usage of `AddrPolicy` in this way when it appears to be possible to just do both. 

> To keep this PR focused: its purpose is to fix the underfilling issue by moving ban and discourage filtering into the selection phase. The current design achieves that with minimal surface area change and without refactoring existing `AddrMan` responsibilities or introducing new abstractions.

If preventing underfilling is the only goal of this PR I would suggest a simpler approach: For example request more addresses than `max_addresses` and then trim at the end. Could be done smartly with using `m_banned.size()` to make sure we only overshoot as much as necessary. But simplest would be to just request with `max_addresses = 0`. The current approach doesn't seem minimal and if there is no plan to use it more broadly it seems like overengineering. At least the conceptual limitations and "policy contract" that you are describing would need to be documented clearly in the code as well.

fjahr commented on src/test/fuzz/addrman.cpp:
```diff
@@ -179,6 +179,27 @@ FUZZ_TARGET(addrman, .init = initialize_addrman)
+    CallOneOf(
+        fuzzed_data_provider,
+        [&] {
+            (void)const_addr_man.GetAddr(max_addresses, max_pct, network, filtered);
```
Why would you call the same line as above again? This line is completely redundant.

waketraindev replied:
oopsie

Bicaru20 commented:
I think @fjahr has a good point. If the main goal is simply to always return 1000 addresses in `GetAddr`, introducing a predicate feels like adding unnecessary complexity. I believe his solution is cleaner. 

> While leaving some nit-ish comments I have realized that if we want to add such a generalized concept for filtering, we should implement the current filtering with it as well. E.g. `network` and `filtered` params go away and we only have `AddrPolicy` as param to `GetAddr`. Then use `AddrPolicy` to implement what `network` and `filtered` were doing. That's a bigger refactor but having this current mix with this redundancy is kind of ugly and I don't really think it's worth it alone.

I agree. If a generalized filtering concept is added, it should replace the current filtering mechanism. This would make it easier to add new filters in the future. However, for the problem trying to be solved, I think it’s complicating things too much.

waketraindev commented:
> I think @fjahr has a good point. If the main goal is simply to always return 1000 addresses in GetAddr, introducing a predicate feels like adding unnecessary complexity. I believe his solution is cleaner.

Regarding the "over-engineering" concerns, the lambda(the predicate) here is inline, concise, and used only once.
It encapsulates a simple banned/discouraged filter applied during selection to ensure max_addresses are correctly returned.
This approach is both functionally necessary and efficient, and does not introduce unnecessary complexity, so I do not consider it over-engineering.

On the other hand, introducing a Builder to craft the policy, include other "generalized" filters, and export AddrInfo would be over-engineering in this context.

The #34162 approach is highly inefficient for a database of 70k+ records, with on-demand calls through RPC with a max_addresses limit, and whitelisted nodes. It is insufficient for my use cases. 

Though both performance and simplicity are in the eyes of the beholder ig.

Thanks for your feedback!

fjahr commented:
> The https://github.com/bitcoin/bitcoin/pull/34162 approach is highly inefficient for a database of 70k+ records, with on-demand calls through RPC with a max_addresses limit, and whitelisted nodes. It is insufficient for my use cases.

If you could be more explicit with what your use-case is then maybe you could convince more people that your approach is really necessary, even when you refuse to use it more broadly. It comes down to if core wants to support your use case bad enough to take on the extra code to maintain, I guess. You have a lot of addresses, ok! But I don't understand how whitelists play into this. Can you elaborate on that?  And why would you do frequent calls to `getnodeaddresses` with a max_addresses limit, couldn't you just increase the limit?

I did some rough high-level benchmarking using the functional test from my PR ([code](https://github.com/fjahr/bitcoin/commit/8f2b63286386b8dbd83981a4c33a4634530323d8)). Due to the bucket collisions I couldn't get beyond 64k addresses, but I guess this is close enough and I didn't try to get to an even higher number. I am using an arbitrary 1k banned addresses and 1k max limit.

This PR (#33663) has the best performance, not surprising:

```
2025-12-30T14:34:41.400316Z TestFramework (INFO): Test that getnodeaddresses returns max count even with banned addresses
2025-12-30T14:34:41.770049Z TestFramework (INFO): Adding addresses to addrman...
2025-12-30T14:36:31.483147Z TestFramework (INFO): Successfully added 62636 addresses
2025-12-30T14:36:31.483466Z TestFramework (INFO): Banning 1000 addresses...
2025-12-30T14:36:32.511395Z TestFramework (INFO): Requesting 1000 addresses...
2025-12-30T14:36:32.523990Z TestFramework (INFO): getnodeaddresses returned 1000 addresses in 0.013s
```

My alternative approach (#34162) is much slower, also not surprising, but at 0.5s it doesn't seem like this would break any workflows I can imagine:

```
2025-12-30T14:18:12.848562Z TestFramework (INFO): Test that getnodeaddresses returns max count even with banned addresses
2025-12-30T14:18:13.218102Z TestFramework (INFO): Adding addresses to addrman...
2025-12-30T14:19:58.335032Z TestFramework (INFO): Successfully added 63471 addresses
2025-12-30T14:19:58.335895Z TestFramework (INFO): Banning 1000 addresses...
2025-12-30T14:19:59.381759Z TestFramework (INFO): Requesting 1000 addresses...
2025-12-30T14:19:59.940406Z TestFramework (INFO): getnodeaddresses returned 1000 addresses in 0.559s
```

As I had said previously, my suggest PR could be tweaked to get a middle ground between performance and simplicity, like for example fetch not all but only 2x max limit addresses to get to comparable performance and this should really work in 99% percent of the cases I can imagine. But sure, it's not as clean as #34162 but performance is comparable to this PR ( [demo code](https://github.com/fjahr/bitcoin/commit/556dcada56aca9136184098e61a07526ab408009)):

```
2025-12-30T14:52:23.292663Z TestFramework (INFO): Test that getnodeaddresses returns max count even with banned addresses
2025-12-30T14:52:23.663001Z TestFramework (INFO): Adding addresses to addrman...
2025-12-30T14:54:09.192166Z TestFramework (INFO): Successfully added 63395 addresses
2025-12-30T14:54:09.193381Z TestFramework (INFO): Banning 1000 addresses...
2025-12-30T14:54:10.218400Z TestFramework (INFO): Requesting 1000 addresses...
2025-12-30T14:54:10.238231Z TestFramework (INFO): getnodeaddresses returned 1000 addresses in 0.020s
2025-12-30T14:54:11.333769Z TestFramework (INFO): Stopping nodes
```

I also learned from playing around with the numbers that the total number of addresses only has minor impact on the performance. The big performance impact is coming from the number of banned addresses. So maybe that's where the discussion on performance (aside from the goals) should focus. I am not sure I have a good grasp on what % of addresses are realistically banned or discouraged. @0xB10C maybe you can weigh in?

waketraindev committed:
addrman, net: filter during selection via AddrPolicy to avoid underfill

Add AddrMan::AddrPolicy, a predicate returning true to exclude an address
during selection. This allows callers to apply custom filtering directly in
AddrMan rather than post-processing results.

Use the new policy in CConnman::GetAddressesUnsafe to exclude banned and
discouraged peers while filling address requests. This fixes underfilled
results observed when banned peers were removed after selection, causing
P2P responses from `GETADDR` to return fewer entries than requested.

Also log the number of filtered entries for debugging and diagnostics.

sipa reviewed (commented):
Concept ACK

But if we're going this direction, I think we could reasonably move all selection logic into to function argument?

I.e. provide just `std::vector<CAddress> CAddrMan::GetAddr(AddrMan::AddrPolicy& policy) const`, and then move the `max_addresses`, `max_pct`, `filtered` arguments into the lambda. It could receive the current number of selected elements + total number of addresses + IsTerribleness in addrman as arguments.
